### DISCUSSION TITLE
Target Audience on Webparts - Hide/show Webparts based on Token evaluation

### DISCUSSION THREAD

**Discussion by edarroudi** [microsoft-search/pnp-modern-search#2047] [Ideas]
I don't know how to call the feature.
What I did and want to discuss is the following solution:

I want to be able to hide a webpart if some parameters are not fulfilled. My main focus is actually some user profile settings or user permissions but I think when we look at it from a general approach it would be better. So I decided to use the TokeService and let the user configure some basic rules. 
I need that for controlling the UI: We'll have users with different roles and it would be nice to control everything inside the same page and not send them to different pages (as 80% of the search layout is the same and only 20% is differen for each user group).

![image](https://user-images.githubusercontent.com/935493/165976738-2a950459-5b6f-456a-8623-680c96f33364.png)

![image](https://user-images.githubusercontent.com/935493/165977834-3aa1e4bd-eda4-49b0-97bc-a32e876b6733.png)
Those are the operators:
![image](https://user-images.githubusercontent.com/935493/165977927-3c631829-c514-4ad6-95c1-fd92acad1fb1.png)

In edit mode
![image](https://user-images.githubusercontent.com/935493/165978085-c0fe470a-1ad6-4966-ad89-cd1cfab43f52.png)

And of course hidden in display mode.

The code itself is really small and i would say not intrusive.

I image it to be useable in the ResultWebpart, FilterWebpart and The verticals, where in the VerticalWebpart we could think of using on each vertical. Probably not that useful for the TextBox Webpart?

The rules are connected through OR so if any of the rules gets evaluated to a true we show the webpart. Of course we always show the webpart when the Target Audience is not set. I thought of using a tree structure configuration that would support AND / OR grouping, but I think that this won't be necessary.

Of course i thought of doing it in the search result webpart in a custom layout with handlebar rules, but the layout is rendered after a search request is sent and that would be bad performace-wise.

Also I think of extending the TokenService with some utitlity for permission checking - I don't think it is implemented yet. Something like {UserPermission.GroupName} or {UserPermission.CanRead.ListX.ItemY} (maybe more simple ones like group checking).

Any comment is appreciated. If you guys give me a go, I'll add the complete code by next week.

**Comment by wobba** [comment]
I think you should post this as a general control to https://pnp.github.io/sp-dev-fx-property-controls/ and then we could add the control to our web parts.

**Reply by Limerick-gh** [reply]
Has this ever been implemented? Or can this be achieved meanwhile differently?
I also like to control either visibility or the search results based on user groups.
Like if/else in the search query with conditions based on security groups.

**Reply by wobba** [reply]
I added this to the script editor web part just now, and could do something similar for search web parts.

https://github.com/pnp/sp-dev-fx-webparts/commit/c68bdfd5946cdc5acb6f751c17cfe53b65db5a89

**Reply by Limerick-gh** [reply]
That sounds great.
So would that mean it could be available by one of the next releases of pnp-modern-search? Or would it require some additional setup to implement it?

**Reply by wobba** [reply]
If I get around to copy the code over, then yes :)

**Reply by KaveSys** [reply]
Hello,
Some update on this? Would be very helpful in Search Filters to filter out some document libraries that should be visible only for a group of users.

**Reply by wobba** [reply]
I don't have time to add it. You can find the code I wrote at https://github.com/pnp/sp-dev-fx-webparts/blob/main/samples/react-script-editor/src/webparts/scriptEditor/ScriptEditorWebPart.ts#L66 and PR it for us :)
